### PR TITLE
Make SDK to reconnect to SENSR when SENSR is re-opened.

### PR DESCRIPTION
**Pull request by KangJH**
## Main changes
- Fix infinite stacking Reconnect() in call stack.
- Make a thread in order to re-connection and do not request it if it's running.

### CODE REVIEW

File: src/sensr_client.cpp [hunk 1]
```diff
@@ -24,32 +24,43 @@ namespace sensr
...
+        size_t reconnect_count_ = Client::kMaxReconnectTrialCount;
+        while (reconnect_count_-- > 0) {
+          for (const auto& listener : listeners_) {
+            UnsubscribeMessageListener(listener);
+          }
+          for (const auto& listener : temp) {
+            SubscribeMessageListener(listener);
+          }
+          bool ret1 = true;
+          bool ret2 = true;
```
Comment (thor-sr) [inline src/sensr_client.cpp:41]: Can we get better names?

File: src/sensr_client.cpp [hunk 2]
```diff
@@ -24,32 +24,43 @@ namespace sensr
...
+              ret2 = false;
+            }
+          }
+
+          if (ret1 && ret2) {
+            INFO_LOG("Reconnected!");
+            break;
+          }
+          std::this_thread::sleep_for(std::chrono::seconds(1));
+          INFO_LOG("Reconnecting...");
```
Comment (thor-sr) [inline src/sensr_client.cpp:58]: Two things
- Why do we need to sleep
- Shouldn't we do `INFO_LOG("Reconnecting...");` before the sleep?

File: src/sensr_client.cpp [hunk 3]
```diff
@@ -24,32 +24,43 @@ namespace sensr
...
+        }
+        is_reconnecting_ = false;
+      });
+      t.detach();
```
Comment (thor-sr) [inline src/sensr_client.cpp:62]: Is it safe to detach this here? What if we get in an endless loop?

File: src/sensr_client.cpp [hunk 4]
```diff
@@ -58,15 +69,27 @@ namespace sensr
...
-        std::bind(&Client::OnResultError, this, std::placeholders::_1));
+         if (std::none_of(listeners_.begin(), listeners_.end(), [](const std::shared_ptr<MessageListener>& listener) {
+          return listener->IsOutputMessageListening();
+        })) {
+          ret = output_endpoint_->Connect(GetProtocol() + "://" + address_ + ":5050", 
+          std::bind(&Client::OnResultMessage, this, std::placeholders::_1),
+          std::bind(&Client::OnResultError, this, std::placeholders::_1));
+        } else {
+          ret = true;
+        }
```
Comment (thor-sr) [inline src/sensr_client.cpp:80]: Can `listeners_` be empty? In that case `std::none_of` returns true. which might not be what we want

File: src/sensr_client.cpp [hunk 5]
```diff
@@ -58,15 +69,27 @@ namespace sensr
     if (std::find(listeners_.begin(), listeners_.end(), listener) == listeners_.end()) {
       // OutputMessage Port
       if (listener->IsOutputMessageListening()) {
-        ret = output_endpoint_->Connect(GetProtocol() + "://" + address_ + ":5050", 
-        std::bind(&Client::OnResultMessage, this, std::placeholders::_1),
-        std::bind(&Client::OnResultError, this, std::placeholders::_1));
+         if (std::none_of(listeners_.begin(), listeners_.end(), [](const std::shared_ptr<MessageListener>& listener) {
+          return listener->IsOutputMessageListening();
```
Comment (thor-sr) [inline src/sensr_client.cpp:73]: Can we not reuse `Client::IsResultListening` here?

File: src/sensr_client.cpp (hunk 1 above)
Comment (KangJH) [reply src/sensr_client.cpp:41]: renamed!
```
bool output_ep_connected = true;
bool point_ep_connected = true;

File: src/sensr_client.cpp (hunk 2 above)
Comment (KangJH) [reply src/sensr_client.cpp:58]: > Why do we need to sleep

Just give some delay to reconnect trial.
Without this, re-connection will be done in few mill-seconds.

> Shouldn't we do INFO_LOG("Reconnecting..."); before the sleep?

Moved it to the very beginning of the loop.

File: src/sensr_client.cpp (hunk 4 above)
Comment (KangJH) [reply src/sensr_client.cpp:80]: Replaced to `if (!IsResultListening()) {` now.

File: src/sensr_client.cpp (hunk 5 above)
Comment (KangJH) [reply src/sensr_client.cpp:73]: Replaced to if (!IsResultListening()) { now.

File: src/sensr_client.cpp (hunk 1 above)
Comment (thor-sr) [reply src/sensr_client.cpp:41]: what does "ep" mean here?

File: src/sensr_client.cpp (hunk 3 above)
Comment (KangJH) [reply src/sensr_client.cpp]: Now, re-connection thread is managed by sdk source.

File: src/sensr_client.cpp (hunk 1 above)
Comment (KangJH) [reply src/sensr_client.cpp]: abbreviation of endpoint

Comment (KangJH) [reply src/sensr_client.cpp]: if you think ep is not good word. then I will just use `output_connected`.
What do you think?

Comment (thor-sr) [reply src/sensr_client.cpp]: yup, that sounds better. I keep reading "ep" as "episode" :rofl:

**Review by thor-sr — APPROVED** [review]
